codex/sales/reservations.nim

```diff
@@ -0,0 +1,314 @@
+## those terms.
+
+import std/typetraits
+
+import pkg/chronos
+import pkg/chronicles
+import pkg/upraises
+import pkg/json_serialization
+import pkg/json_serialization/std/options
+import pkg/stint
+import pkg/nimcrypto
+
+push: {.upraises: [].}
+
+import pkg/datastore
+import pkg/stew/byteutils
+import ../stores
+import ../namespaces
+import ../contracts/requests
+
+export requests
+
+type
+  AvailabilityId* = distinct array[32, byte]
+  Availability* = object
+    id*: AvailabilityId
+    size*: UInt256
+    duration*: UInt256
+    minPrice*: UInt256
+    slotId*: ?SlotId
+  Reservations* = ref object
+    repo: RepoStore
+  GetNext* = proc(): Future[?Availability] {.upraises: [], gcsafe, closure.}
+  AvailabilityIter* = ref object
+    finished*: bool
+    next*: GetNext
+  AvailabilityError* = object of CodexError
+  AvailabilityNotExistsError* = object of AvailabilityError
+  AvailabilityAlreadyExistsError* = object of AvailabilityError
+  AvailabilityReserveFailedError* = object of AvailabilityError
+  AvailabilityReleaseFailedError* = object of AvailabilityError
+  AvailabilityDeleteFailedError* = object of AvailabilityError
+  AvailabilityPutFailedError* = object of AvailabilityError
+  AvailabilityGetFailedError* = object of AvailabilityError
+
+const
+  SalesKey = (CodexMetaKey / "sales").tryGet # TODO: move to sales module
+  ReservationsKey = (SalesKey / "reservations").tryGet
+
+proc new*(
+  T: type Reservations,
+  repo: RepoStore): Reservations =
+
+  T(repo: repo)
+
+proc init*(
+  _: type Availability,
+  size: UInt256,
+  duration: UInt256,
+  minPrice: UInt256): Availability =
+
+  var id: array[32, byte]
+  doAssert randomBytes(id) == 32
+  Availability(id: AvailabilityId(id), size: size, duration: duration, minPrice: minPrice)
+
+func toArray*(id: AvailabilityId): array[32, byte] =
+  array[32, byte](id)
+
+proc `==`*(x, y: AvailabilityId): bool {.borrow.}
+proc `==`*(x, y: Availability): bool =
+  x.id == y.id and
+  x.size == y.size and
+  x.duration == y.duration and
+  x.minPrice == y.minPrice
+
+proc toErr[E1: ref CatchableError, E2: AvailabilityError](
+  e1: E1,
+  _: type E2,
+  msg: string = "see inner exception"): ref E2 =
+
+  return newException(E2, msg, e1)
+
+proc writeValue*(
+  writer: var JsonWriter,
+  value: SlotId | AvailabilityId) {.upraises:[IOError].} =
+
+  mixin writeValue
+  writer.writeValue value.toArray
+
+proc readValue*[T: SlotId | AvailabilityId](
+  reader: var JsonReader,
+  value: var T) {.upraises: [SerializationError, IOError].} =
+
+  mixin readValue
+  value = T reader.readValue(T.distinctBase)
+
+func used*(availability: Availability): bool =
+  availability.slotId.isSome
+
+func key(id: AvailabilityId): ?!Key =
+  (ReservationsKey / id.toArray.toHex)
+
+func key*(availability: Availability): ?!Key =
+  return availability.id.key
+
+func available*(self: Reservations): uint =
+  return self.repo.quotaMaxBytes - self.repo.totalUsed
```

Comment:
This should probably be moved to the repo.

Reply:
done